modules/placement-driver/src/test/java/org/apache/ignite/internal/placementdriver/PlacementDriverTest.java

```diff
@@ -254,6 +268,55 @@ public void testAwaitPrimaryReplicaBeforeIntervalAfterPublishing() throws Except
         assertEquals(LEASE_FROM_1_TO_15_000.getExpirationTime(), primaryReplicaFuture.get().getExpirationTime());
     }
 
+    /**
+     * Test steps.
+     * <ol>
+     *     <li>Publish primary replica for an interval [1, 5].</li>
+     *     <li>Leaseholder of the primary replica is offline</li>
+     *     <li>Await primary replica for timestamp 3, this shouldn't complete instantly.</li>
+     *     <li>Publish a new lease for interval [5, 10] and another leaseholder node that is online.</li>
+     *     <li>Assert that primary waiter is completed and the new lease belongs to the online node.</li>
+     * </ol>
+     */
+    @Test
+    public void testAwaitCurrentPrimaryIsOffline() {
+        // Publish primary replica for an interval [1, 5].
+        publishLease(LEASE_FROM_1_TO_5_000);
+
+        // Cluster node resolver will should that the current leaseholder is offline.
+        leaseholder = null;
+
+        CompletableFuture<ReplicaMeta> primaryReplicaFuture =
+                placementDriver.awaitPrimaryReplica(GROUP_1, AWAIT_TIME_3_000, AWAIT_PRIMARY_REPLICA_TIMEOUT, SECONDS);
+
+        assertFalse(primaryReplicaFuture.isDone());
+
+        String newLeaseholder = "newLeaseholder";
+        Lease newLease = new Lease(
+                newLeaseholder,
+                newLeaseholder,
+                new HybridTimestamp(5_001, 0),
```

Comment:
I'd also add two more tests

- newLease.startTime =  old.expirationTimestamp + some serious delta (e.g. 1000) in order to emulate gaps for new replica election. Basically that will cover non exact match of tick and 5000 + 1.
- future timeout verification test.

Reply:
added